src/vhost.erl

```diff
@@ -0,0 +1,167 @@
+%% The contents of this file are subject to the Mozilla Public License
+%% Version 1.1 (the "License"); you may not use this file except in
+%% compliance with the License. You may obtain a copy of the License
+%% at https://www.mozilla.org/MPL/
+%%
+%% Software distributed under the License is distributed on an "AS IS"
+%% basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
+%% the License for the specific language governing rights and
+%% limitations under the License.
+%%
+%% The Original Code is RabbitMQ.
+%%
+%% The Initial Developer of the Original Code is GoPivotal, Inc.
+%% Copyright (c) 2018-2019 Pivotal Software, Inc.  All rights reserved.
+%%
+
+-module(vhost).
+
+-include_lib("rabbit_common/include/rabbit.hrl").
+-include("vhost.hrl").
+
+-export([
+  new/2,
+  new/3,
+  fields/0,
+  fields/1,
+  info_keys/0,
+  record_version_to_use/0,
+  upgrade/1,
+  upgrade_to/2,
+  pattern_match_all/0,
+  get_name/1,
+  get_limits/1,
+  get_metadata/1,
+  get_description/1,
+  get_tags/1,
+  set_limits/2
+]).
+
+-record(vhost, {
+          %% name as a binary
+          virtual_host,
+          %% proplist of limits configured, if any
+          limits,
+          metadata}).
+
+-define(record_version, vhost_v2).
+
+-type(name() :: binary()).
+
+-type vhost() :: vhost_v1:vhost_v1() | vhost_v2().
+
+-type vhost_v2() :: #vhost{
+                          virtual_host :: rabbit_types:vhost(),
+                          limits :: list(),
+                          metadata :: map() | undefined
```

Comment:
You can define it further by specifyfing the type of `description` and `tags` and mark them as optional.

I would define a dedicated type for `metadata` to be able to export it and use it in this file.